app/scripts/directives/oscKeyValues.js

```diff
@@ -162,10 +117,112 @@ angular.module("openshiftConsole")
         readonlyKeys: "@",
         keyValidationTooltip: "@",
         valueValidationTooltip: "@",
-        preventEmpty: "=?",
+        preventEmpty: "=?"
       },
       controller: function($scope){
+        var focusElem;
+        var added = {};
+        var isUncommitted = function() {
+          return (!!$scope.key) || (!!$scope.value);
+        };
+        var checkCommitted = function() {
+          if(isUncommitted()) {
+            $scope.showCommmitWarning = true;
+          } else {
+            $scope.showCommmitWarning = false;
+          }
+        };
+        // checks if the key,value inputs have any text value.
+        // if so, sets the form name="clean" to an 'invalid' state, which will
+        // invalidate any parent form up the chain.  This should result in an
+        // inability to submit that form until the user commits the new key-value pair.
+        var isClean = _.debounce(function() {
+          $scope.$applyAsync(function() {
+            if(!!$scope.key) {
+              $scope.clean.isClean.$setValidity('isClean', false);
+            } else if(!!$scope.value) {
+              $scope.clean.isClean.$setValidity('isClean', false);
+            } else {
+              $scope.clean.isClean.$setValidity('isClean', true);
+            }
+          });
+        }, 100);
+
+        var onBlur = function(nodes) {
+          return _.debounce(function(evt) {
```

Comment:
Not really, I will eliminate it and update the `$scope.$apply` to a `$scope.$applyAsync` to avoid potential digest errors.